[FLINK-28474][checkpoint] Fix the bug ChannelStateWriteResult might not fail after checkpoint abort

1996fanrui opened the pull request:
## What is the purpose of the change

Simplify the ChannelStateWriterImpl due to concurrent unaligned checkpoint isn't supported.

## Brief change log

- Simplify the ChannelStateWriterImpl due to concurrent unaligned checkpoint isn't supported.
- If the checkpointId is aborted, the checkpoint will be ignored within `ChannelStateWriter#start`.


## Verifying this change

This change added tests and can be verified as follows:

  - *ChannelStateWriterImplTest#testStartMultipleTimes()*
  - *ChannelStateWriterImplTest#testBuffersRecycledOnAbortedCheckpoint()*
  - *ChannelStateWriterImplTest#testOldCheckpointIsAborted()*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

1996fanrui commented:
@flinkbot run azure

pnowojski commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorTest.java:
```diff
@@ -568,6 +572,53 @@ public void snapshotState(
 
+    @Test
+    public void testChannelStateWriteResultLeakAndNotFailAfterCheckpointAborted() throws Exception {
+        MockEnvironment mockEnvironment = MockEnvironment.builder().build();
```
wrap with `try-with-resource`?

1996fanrui replied:
done

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -316,6 +316,10 @@ public void checkpointState(
             // broadcast cancel checkpoint marker to avoid downstream back-pressure due to
             // checkpoint barrier align.
             operatorChain.broadcastEvent(new CancelCheckpointMarker(metadata.getCheckpointId()));
+            channelStateWriter.abort(
+                    metadata.getCheckpointId(),
+                    new CancellationException("checkpoint aborted via notification"),
+                    true);
```
I think you can not relay on this code to be executed. `checkAndClearAbortedStatus` is using best effort `abortedCheckpointIds` set, which can be for example pruned if it grows too large. So if FLINK-26803 doesn't work without this fix, we have to deal with it differently.

1996fanrui replied:
Thanks for your clarification and reminder, this is really a bug. 

I have updated this PR that fixed two other places where remove aborted checkpoint.


> What exact issue was this bug causing in [FLINK-26803](https://issues.apache.org/jira/browse/FLINK-26803) and what were the symptoms?

The bug will affect the new feature [FLINK-26803](https://issues.apache.org/jira/browse/FLINK-26803), because the channel state file can be closed only after the Checkpoints of all tasks of the shared file are complete or abort. So when the checkpoint of some tasks fails, if abort is not called, the file cannot be closed and all tasks sharing the file cannot execute `inputChannelStateHandles.completeExceptionally(e);` and `resultSubpartitionStateHandles.completeExceptionally(e);` , these AsyncCheckpointRunnable of shared tasks will wait forever.

If the channel state file is not shared between tasks, the current task will not create an AsyncCheckpointRunnable, so the current task and other tasks will not be affected.

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
@flinkbot run azure

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:
```diff
@@ -60,89 +63,78 @@
     private static final Logger LOG = LoggerFactory.getLogger(ChannelStateWriterImpl.class);
-    private static final int DEFAULT_MAX_CHECKPOINTS =
-            1000; // includes max-concurrent-checkpoints + checkpoints to be aborted (scheduled via
-    // mailbox)
```
Hi @pnowojski 

From the comment, we can get that the `ConcurrentMap<Long, ChannelStateWriteResult> results;` saves ongoingCheckpoints and abortedCheckpoints. 

But the checkpoints that were aborted before start were not saved. We want aborted checkpoints to ignore start. So I introduced the ` NavigableSet<Long> abortedCheckpointIds;` to save all aborted checkpointIds. 

- If checkpointId is aborted before start, it will be saved into this set. 
- When the start is called, if checkpointId in this set, the start should be ignored.

And introduced the `long ongoingCheckpointId;` to save the ongoing checkpointId. Because the concurrent unaligned checkpoint isn't supported, so the `long` is enough. And checkpoints with checkpointId less than `ongoingCheckpointId` should be aborted. So I remove these checkpointIds when the ongoingCheckpointId is updated. 

What do you think?

1996fanrui replied:
Hi @pnowojski , thanks for your review.

The `checkpointId < nextExpectedCheckpointId` may happen. For example, `AlternatingWaitingForFirstBarrierUnaligned#barrierReceived` don't check whether checkpointId is aborted, if received a barrier, it will call the ChannelStateWriter.start().

So, we should use `ongoingCheckpointId`, and just update it within `ChannelStateWriterImpl#start`

We should ignore checkpoint when the checkpointId is aborted, so we  need the `NavigableSet<Long> abortedCheckpointIds;`.

pnowojski replied:
Thanks for the explanation. But in that case, why do we need `abortedCheckpointIds` set? Couldn't we have just `long ongoingCheckpointId` and `long nextExpectedCheckpointId`?

1996fanrui replied:
Hi @pnowojski , great idea! 

Updated, but I used maxAbortedCheckpointId instead of nextExpectedCheckpointId. It should be clearer.

pnowojski reviewed (commented):
Thanks for the changes, please see a couple of my comments.

1996fanrui commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:
```diff
@@ -60,89 +63,78 @@
+    // ongoingCheckpointId should be aborted due to concurrent unaligned checkpoint is currently not
+    // supported.
+    @GuardedBy("lock")
+    private ChannelStateWriteResult result;
```
Removed

pnowojski replied:
Ops, sorry. I've mislead you. Now I remember it was supposed to be task owned class, but as a result of some discussions it ended up being used from the Netty thread (`addInputData` called from `ChannelStatePersister#maybePersist`). That's also why we needed `ConcurrentMap` here before.

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
@flinkbot run azure

1996fanrui commented:
@flinkbot run azure

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:
```diff
@@ -60,89 +63,73 @@
+                return;
+            }
+            if (result != null) {
+                result.fail(new CancellationException("Cancel old checkpoint."));
```
I think it would be safer to fail the futures outside of this lock. We don't know what other locks will be acquired from the actions that are connected to those futures.

So something like:

```
@Nullable ChannelStatWriteResult toFail = null;

synchronized (lock) { 
  ...
  toFail = result;
  ...
}

if (toFail != null) {
  toFail.fail();
}
```
?

Or maybe for the sake of being consistent use the same pattern as in other places?

```
 enqueue(ChannelStateWriteRequest.abort(...));
```

as it _seems_ to achieve a similar thing?

I'm not sure which of this option is better here.

1996fanrui replied:
@pnowojski , thanks for your reminder.

It's a good advice. The fail will close stream, it may be slow, so it should hold the lock.

I prefer `enqueue(ChannelStateWriteRequest.abort(...));`, if there is any problem with fail it should not affect the task thread

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -120,15 +141,28 @@ public void testAbortIgnoresMissing() throws Exception {
+        assertNull(writer.getWriteResult(1));
+        assertFalse(writer.getWriteResult(2).isDone());
+
+        assertWriteResultIsCompletedExceptionally(result1);
```
Can you add assertion that the exception is `CheckpointException(CheckpointFailureReason#CHECKPOINT_DECLINED_SUBSUMED)`?

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -121,15 +141,28 @@ public void testAbortIgnoresMissing() throws Exception {
     }
 
+    @Test
+    public void testOldCheckpointIsAborted() {
```
Aren't we missing a test coverage for the main point of this bug? That the following sequence:
```
callAbort(1, ...);
callStart(1, ...);
```
also leads to a completed `ChannelStateWriteResult`? Previously that would create an uncompleted `ChannelStateWriteResult`, that was causing a deadlocks in FLINK-26803 right?

pnowojski commented:
I've started a benchmark request to check the impact of replacing `ConcurrentHashMap` with a lock.

http://codespeed.dak8s.net:8080/job/flink-benchmark-request/171/

Also FYI @rkhachatryan, I've already reviewed this code, but you might be interested in this issue.

1996fanrui commented:
> Thanks @1996fanrui, I think the code looks mostly good. I've started a benchmark request to check the impact of replacing `ConcurrentHashMap` with a lock. I don't expect it to be an issue, but let's check that.
> 
> http://codespeed.dak8s.net:8080/job/flink-benchmark-request/171/
> 
> Also FYI @rkhachatryan, I've already reviewed this code, but you might be interested in this issue.

Checkpoint is a relatively low-frequency behavior. My understanding is that this change should have little or no impact on data processing.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java:
```diff
@@ -157,6 +151,14 @@ public void addInputData(
                 checkpointId,
                 info,
                 startSeqNum);
+        if (isCheckpointSubsumedOrAborted(checkpointId)) {
+            LOG.debug(
+                    "The checkpoint {} of task {} has been aborted, so don't addInputData.",
+                    checkpointId,
+                    taskName);
+            closeBuffers(iterator);
+            return;
+        }
         enqueue(write(checkpointId, info, iterator), false);
```
What will happen if there is a race condition between this `addInputData` call from the netty threads, and for example `abort` call?

1. (Netty thread) calling `addInputData`
2. (Netty thread) checking `isCheckpointSubsumedOrAborted`, method returns false
3. (Netty thread) freezes for a 1ms
4. (Task thread) calls `abort()`
5. (Netty thread) wakes up and finishes `addInputData` method by calling `enqueue(write(checkpointId, info, iterator), false);`

Will this lead to a resource leak? If not, then why do we need `if (isCheckpointSubsumedOrAborted(checkpointId))` check here in the first place? `isCheckpointSubsumedOrAborted` is quite costly due to lock acquisition after all.

pnowojski replied:
~this issue is no longer relevant~ we decided to go with another approach

pnowojski reviewed (commented):
I've checked the benchmark results and didn't notice any performance regression. I have one more comment however.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImplTest.java:
```diff
@@ -120,15 +146,55 @@ public void testAbortIgnoresMissing() throws Exception {
+        writer.open();
+        callAbort(writer);
+
+        // It should be started.
```
nit: this comment is a bit miss leading. `it should be started` -> `start after abort should be ignored`?

1996fanrui commented:
Hi @pnowojski , I updated this PR based on the first version. Please help take a look in your free time, thanks a lot~

pnowojski reviewed (commented):
Thanks @1996fanrui for the update.  I think this version looks much simpler compared to the previous. And sorry again for the detour.

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -209,6 +213,7 @@ class SubtaskCheckpointCoordinatorImpl implements SubtaskCheckpointCoordinator {
         this.prepareInputSnapshot = prepareInputSnapshot;
         this.abortedCheckpointIds =
                 createAbortedCheckpointSetWithLimitSize(maxRecordAbortedCheckpoints);
+        this.maxRecordAbortedCheckpoints = maxRecordAbortedCheckpoints;
```
Thanks for your review.

Actually, I have a question: why need the `SubtaskCheckpointCoordinatorImpl#DEFAULT_MAX_RECORD_ABORTED_CHECKPOINTS`? Why don't use the `max concurrent checkpoints` here?

pnowojski replied:
> Could we always use the max concurrent checkpoints and remove the SubtaskCheckpointCoordinatorImpl#DEFAULT_MAX_RECORD_ABORTED_CHECKPOINTS?

I think so.

> why need the SubtaskCheckpointCoordinatorImpl#DEFAULT_MAX_RECORD_ABORTED_CHECKPOINTS? Why don't use the max concurrent checkpoints here?

Probably because when it was being implemented by Myasuka (Yun Tang) and reviewed by me 2 years ago, we haven't figured out this trick with using `max concurrent checkpoints` :(

1996fanrui replied:
If so, I can fix it. I should fix it in this pr and using a new commit, right?

pnowojski replied:
👍

1996fanrui commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -312,10 +317,15 @@ public void checkpointState(
         // if necessary.
         lastCheckpointId = metadata.getCheckpointId();
-        if (checkAndClearAbortedStatus(metadata.getCheckpointId())) {
+        if (checkAndClearAbortedStatus(metadata.getCheckpointId())
+                || isAbortedAndRemovedCheckpointId(metadata.getCheckpointId())) {
```
If removed, the `checkAndClearAbortedStatus` will return true. I don't know whether the `isAbortedAndRemovedCheckpointId` will affects `checkAndClearAbortedStatus` in the future. 

Currently, it's ok. The return value just be used here.

1996fanrui commented:
> Thanks @1996fanrui for the update. I think this version looks much simpler compared to the previous. And sorry again for the detour.
> 
> I've left a couple of smaller comments.

Hi @pnowojski , thanks for your review, I have addressed all comments.

pnowojski commented:
Thanks @1996fanrui for the update LGTM % last final minor issue that I think came back after reverting to the earlier version:
https://github.com/apache/flink/pull/20233#discussion_r988921400

1996fanrui commented:
@flinkbot run azure

pnowojski commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorTest.java:
```diff
@@ -569,6 +571,58 @@ public void snapshotState(
+                    false,
+                    () -> true);
+            assertNull(writer.getWriteResult(checkpointId));
+            TimeUnit.MILLISECONDS.sleep(10);
```
One more thing. Can we get rid of this sleep? It's very likely it will be failing from time to time in the azure CI.

It looks like we can replace it with
```
writeResult.get();
```
where

```
@VisibleForTesting
public void ChannelStateWriter#get() {
  inputChannelStateHandles.get();
  resultSubpartitionStateHandles.get();
}
```

1996fanrui replied:
Thanks for your reminder. The unit test will fail when using the `ChannelStateWriter#get()` directly.

`writeResult.getInputChannelStateHandles()` and `writeResult.getResultSubpartitionStateHandles()` are completed exceptionally, so the get will throw exception.

I replace it with `ChannelStateWriter#waitForDone()`, it just wait for `inputChannelStateHandles` and `resultSubpartitionStateHandles` are done, and ignore any exception.

```
@VisibleForTesting
public void waitForDone() {
    try {
        inputChannelStateHandles.get();
    } catch (Throwable ignored) {
    }
    try {
        resultSubpartitionStateHandles.get();
    } catch (Throwable ignored) {
    }
}
```
